### PR TITLE
use nvme for nova ephemeral disk

### PR DESCRIPTION
**Pull request by venkataanil**
This will mount nvme device on compute node and will use that as
nova ephemeral storage so that instances are booted on nvme device.

### CODE REVIEW

File: templates/firstboot-nvme.yaml.j2 [hunk 1]
```diff
@@ -33,6 +33,11 @@ resources:
             if [[ $(hostname) == *$FORMAT* ]] ; then
               sed "s/^\(GRUB_CMDLINE_LINUX=\".*\)\"/\1 $KERNEL_ARGS\"/g" -i /etc/default/grub ;
               grub2-mkconfig -o /etc/grub2.cfg
+              mkdir -p /var/lib/nova
+              mkfs -t xfs -f /dev/nvme0n1
+              mount /dev/nvme0n1 /var/lib/nova
+              chmod a+w /var/lib/nova
+              echo UUID=`sudo blkid -s UUID -o value /dev/nvme0n1` /var/lib/nova xfs discard,defaults,nofail 0 2 | sudo tee -a /etc/fstab
```
Comment (smalleni) [inline templates/firstboot-nvme.yaml.j2]: Do you plan to hardcode /dev/nvme0n1 here?

File: group_vars/all.yml [hunk 2]
```diff
@@ -167,6 +167,11 @@ shift_stack: false
 #installs browbeat, if this variable is not defined or set true
 #skips browbeat installation, only if the variable set false
 browbeat: true
+
+# define mount_nvme to mount /var/lib/nova on nvme device
+# to use nvme storage for nova instances as ephemeral disk
+#mount_nvme: true
```
Comment (smalleni) [inline group_vars/all.yml:174]: I recommend mount_nvme to be a dictionary with one of the keys being the path of the nvme device (/dev/nvmen01..). Similarly for the passthrough stuff, we should change the name from `nvme` to something like `passthrough_nvme` to distinguish

File: templates/firstboot-nvme.yaml.j2 (hunk 1 above)
Comment (smalleni) [inline templates/firstboot-nvme.yaml.j2]: Do you plan to keep the nvme path hardcoded? Would be good to parameterize inside of the `mount_nvme` dictionary.

File: overcloud.yml [hunk 3]
```diff
@@ -26,7 +26,7 @@
         template:
           src: "firstboot-nvme.yaml.j2"
           dest: "/home/stack/firstboot-nvme.yaml"
-        when: nvme is defined
+        when: nvme is defined or mount_nvme is defined
```
Comment (smalleni) [inline overcloud.yml]: The template also have some fields like ComputeKernelArgs that are only needed for the passthrough case, so we should have a condition within the template.

File: templates/firstboot-nvme.yaml.j2 [hunk 4]
```diff
@@ -33,6 +33,11 @@ resources:
             if [[ $(hostname) == *$FORMAT* ]] ; then
               sed "s/^\(GRUB_CMDLINE_LINUX=\".*\)\"/\1 $KERNEL_ARGS\"/g" -i /etc/default/grub ;
               grub2-mkconfig -o /etc/grub2.cfg
+              mkdir -p /var/lib/nova
+              mkfs -t xfs -f /dev/nvme0n1
+              mount /dev/nvme0n1 /var/lib/nova
+              chmod a+w /var/lib/nova
```
Comment (smalleni) [inline templates/firstboot-nvme.yaml.j2]: Also you can't do this mounting if you want to passthrough nvme as the device won't be available on the hypervisor. Looks like you need a condition within the template to do this only for the case where you mount nvme.

**Review by smalleni — CHANGES_REQUESTED** [review]

File: templates/firstboot-nvme.yaml.j2 (hunk 4 above)
Comment (smalleni) [reply templates/firstboot-nvme.yaml.j2]: Is the file system xfs even on OSP 13 ? I rememberthat the default is different between OSP 13 and OSP 16, you might want to check.

File: templates/firstboot-nvme.yaml.j2 (hunk 1 above)
Comment (venkataanil) [reply templates/firstboot-nvme.yaml.j2]: nvme0n1 naming is used for nvme device. We can use only one nvme device for nova instances.
This hard coding will be a problem only when we have multiple nvme devices and other nvme devices are of bigger size than the first nvme device.

File: templates/firstboot-nvme.yaml.j2 (hunk 4 above)
Comment (venkataanil) [reply templates/firstboot-nvme.yaml.j2]: Agree. As we know, NVME passthrough is introduced for OCP. As OCP can't use it (OCP IPI limitation), thought of removing NVME passthrough in later patches.

**Comment by venkataanil** [comment]
Sai, I agree for all your suggestions. I will push a new patchset.
Masco, you can still test with this first patchset.

File: templates/firstboot-nvme.yaml.j2 (hunk 4 above)
Comment (venkataanil) [reply templates/firstboot-nvme.yaml.j2]: > Is the file system xfs even on OSP 13 ? I rememberthat the default is different between OSP 13 and OSP 16, you might want to check.

Both OSP13 and OSP16.1 are using xfs file systems.

**Comment by masco** [comment]
i am deploying with nvme disabled

**Comment by links84** [comment]
I was hoping if jetpack finds more nvme disk, it should create and lvm with all nvme disk and mount it.

**Comment by venkataanil** [comment]
> I was hoping if jetpack finds more nvme disk, it should create and lvm with all nvme disk and mount it.

Can you please elaborate?

**Comment by links84** [comment]
> > I was hoping if jetpack finds more nvme disk, it should create and lvm with all nvme disk and mount it.
> 
> Can you please elaborate?

We have nodes with 4 nvme disks too. In future if we have higher storage requirements per compute, we can use such nodes and make sure all 4 nvme disks are added to lvm and mounted .
We can use stripe = numer of disk to improve performance.
lvcreate .. -i <number of disks>  <VG name>

File: group_vars/all.yml (hunk 2 above)
Comment (venkataanil) [reply group_vars/all.yml:174]: we can think about using as dict later when we get other requirements

Comment (masco) [reply group_vars/all.yml:174]: @venkataanil can you add these variables in CI files too.

**Comment by masco** [comment]
rerun tests

**Pull request merged**